### PR TITLE
IVF-PQ: store cluster data in individual lists and reduce templates

### PR DESCRIPTION
**Pull request by achirkin**
Refactor `ivf_pq::index` to keep the cluster data in separate lists/buffers and reduce the number of template instantiations where possible.

- Breaking change: the public structure `ivf_pq::index` is changed.
- Partially addresses #1201: removing the `mdarray<IdxT, ...> list_offsets` member makes it possible to tweak the search kernel to remove the `IdxT` template parameter altogether and stick to `uint32_t` _sample_ indices, which are then backtraced to the database indices (`IdxT`) during the post-processing stage.
- Partially addresses #1170
- Improves the index extending time:
  - No need to calculate offsets, reorder clusters, etc.
  - The individual lists are shared between multiple versions of the index and amortize the allocation/copy costs.
- Small addition to the tests: split serialize/deserialize tests in a separate test cases to allow checking the index state in all possible scenarios (search after: building new, extending, deserializing).

### CODE REVIEW

**Comment by achirkin** [comment]
TODO:
- [ ] Refactor the index `extend` code
- [ ] Test the impact on compile times
- [ ] Test the impact on search runtime

**Comment by cjnolet** [comment]
Hey @achirkin, I did some cleanup and improvments to the ivf-pq build by consolidating the templated types between benchmarks and tests. Can you give that PR a reivew please so we can get those changes merged into this PR as well?

**Comment by achirkin** [comment]
NB: I've reverted some of your changes to ivf-pq, because they don't make sense anymore due to my changes to the logic. I did my best to keep the other changes, but may have reverted some of them due to the mayhem caused by the git merge :)

Note, I've left the 32-bit indexed tests in ivf-pq on purpose. The ivf_pq_similarity_kernel (which has the most instances by a large margin) doesn't depend on the indexing type anymore, so build times of the tests shouldn't add too much overhead. Yet, I believe it's important to keep testing the intricate logic of converting `IdxT<->uint32_t` that happens in pre/post processing steps of the search. Otherwise, the other index types will surely stop compile after a few more PRs.

File: cpp/test/CMakeLists.txt [hunk 1]
```diff
@@ -244,6 +244,8 @@ if(BUILD_TESTS)
     test/neighbors/ann_ivf_flat/test_float_int64_t.cu
     test/neighbors/ann_ivf_flat/test_int8_t_int64_t.cu
     test/neighbors/ann_ivf_flat/test_uint8_t_int64_t.cu
+    test/neighbors/ann_ivf_pq/test_float_int64_t.cu
```
Comment (cjnolet) [inline cpp/test/CMakeLists.txt:247]: Realistically, do we expect to see users using `int64_t` index types when `uint64_t` is already available? Or am I misunderstanding the intention of this index param type? I propose we standardize upon one of them and use them consistently across the APIs and documentation. I'd personally prefer to use `int64_t` over `uint64_t` because that's what I believe faiss has recently switched (back) to using across the board ([ref](https://github.com/facebookresearch/faiss/commit/43d3dcf06184a6351d1fb0e8277c1238c71476b6)). I think we should also consider standardizing on these types across raft in general. I've been considering ways to document and communicate the specialized types to users and it would be great if we could send a consistent message. 

While these additional instantiations may not seem like a lot of additional time on a single architecture, please consider that CI is building for 6 different architectures and `sccache` has been, I fear` sending us a false message as to the amount of time it actually takes to build the codebase end-to-end. An additional instantiation which makes a test take 5 minutes longer to compile for a single architecture could add 20+ minutes to CI. 5 different instantiations which make this assumption could extend CI by over an hour. Right now the challenge is that RAFT is being statically built into both cugraph and cuml's wheels at the moment so these additional compile times extend further than just raft's CI.

**Review by cjnolet — CHANGES_REQUESTED** [review]
I'm definiely excited to break the IVFs out into individual lists. I'm also excited to reduce more templates. I think we need to have a broader conversation about how we might be able to condoliate the types used across RAFT into a more consistent set. We've made attempts to do this before for specific namespaces but I don't think we've really discussed how we can do this across the codebase.

**Comment by cjnolet** [comment]
I'm cross-referencing this comment because it also relates to the IVF* build times: https://github.com/rapidsai/raft/pull/1254#issuecomment-1423043221.

File: cpp/test/CMakeLists.txt (hunk 1 above)
Comment (achirkin) [reply cpp/test/CMakeLists.txt:249]: I don't particularly want test `int64_t` specifically, but I still would like to keep at least one extra test with another index type besides `uint64_t` (at the moment I have two here). Maybe I'd remove this, but keep `uint32_t` version?
TL;DR

I totally agree with our move to stick to one index type, and I also slightly prefer a signed 64 bit type, because it's a little bit safer (e.g. against underflows in reverse loops).

Something I hope but haven't thoroughly tested yet, is that setting a different index type should add a negligible extra compile time compared to the similarity kernel (because the index type is converted to `uint32_t` anyway). Also I surely don't want to add these extra instances to the distance/specializations library, but keep it only in tests (so that it does not affect downstream projects).

At the same time, I appreciate the raft's flexible interface providing value/index template parameters. That allows the downstream projects like faiss to stick to their indexing types and avoid doing extra data conversion passes on every raft invocation and simplify the integration in general. I like that our model of flexible templates + limited specializations allows both alternatives, either fast compile times using raft shared library or flexible instances and compilation on the downstream project side.
The problem we have already encountered in raft in the past, is that we provide flexible templates but test them  with only one template instance. Then, some of the functions don't even compile with another instance, because somewhere in the middle of the code we hardcode something like `rmm::device_uvector<uint64_t> tmp_buffer` instead of `rmm::device_uvector<IdxT>` by accident.

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh [hunk 2]
```diff
@@ -16,7 +16,7 @@
 
 #pragma once
 
-#include "ann_utils.cuh"
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:19]: Let's move it a bit later together with ivf-flat (I didn't intend to do this much refactoring in this PR to avoid hiding the changes in ivf-pq logic behind all this).

**Comment by achirkin** [comment]
Update: I found out today that our `ivfpq_compute_similarity` instantiations were not picked up by the `ivf_pq::search` instantiations leading to longer compile times of the latter. The fix is almost there, I'll push it tomorrow.

**Comment by cjnolet** [comment]
@achirkin that could be quite a savings then, I hope!

File: cpp/include/raft/neighbors/ivf_pq_types.hpp [hunk 3]
```diff
@@ -139,6 +142,80 @@ constexpr static uint32_t kIndexGroupSize = 32;
...
+      RAFT_FAIL(
+        "ivf-pq: failed to allocate a big enough index list to hold all data "
+        "(requested size: %zu records, selected capacity: %zu records). "
+        "Allocator exception: %s",
+        size_t(size),
+        size_t(capacity),
+        e.what());
+    }
+    // Fill the index buffer with a pre-defined marker for easier debugging
+    thrust::fill_n(
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/ivf_pq_types.hpp:214]: Unfortunately, I had to revert this change, because `raft::matrix::fill` is in `.cuh` file and contains cuda code, whereas `ivf_pq_types.hpp` must not.

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks Artem for this PR, it is great to see this improvement in the index structure! Please see my first batch of comments related to the `build` method.

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh [hunk 4]
```diff
@@ -1168,7 +1105,21 @@ auto extend(raft::device_resources const& handle,
...
+ * See raft::spatial::knn::ivf_pq::extend docs.
+ */
+template <typename T, typename IdxT>
+auto extend(raft::device_resources const& handle,
+            const index<IdxT>& orig_index,
+            const T* new_vectors,
+            const IdxT* new_indices,
+            IdxT n_rows) -> index<IdxT>
+{
+  auto ext_index = clone(handle, orig_index);
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:1121]: This is a part of the already exposed public api. We have two versions of the `extend` function. One updates the given index, the other create a new one. This one is the latter; the orig_index comes as `const` and cannot be modified.

NB: in the previous version, we always created new index internally, and just replaced the original afterwards. In the new version we can go either way, but also the new index in the latter version can share some of the data with the previous index.

File: cpp/include/raft/neighbors/ivf_pq_types.hpp [hunk 5]
```diff
@@ -139,6 +142,80 @@ constexpr static uint32_t kIndexGroupSize = 32;
...
+
+/** The data for a single list (cluster). */
+template <typename IdxT, typename SizeT = uint32_t>
+struct list_data {
+  /** PQ-encoded interleaved data. */
+  device_mdarray<uint8_t, list_extents<SizeT>, row_major> data;
+  /** Source indices. */
+  device_mdarray<IdxT, extent_1d<SizeT>, row_major> indices;
+  /** The actual size of the content. */
+  std::atomic<SizeT> size;
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/ivf_pq_types.hpp:194]: The concept is that `list_data` is safe to grow in a concurrent environment. That is, if a user extends multiple databases from one common database during their experimenting, these multiple versions may still refer to the same `list_data` for a cluster, even if one of them extended the cluster. The `size` then refers to the maximum size of this cluster among the databases, in which this cluster is shared. It is atomic to make sure we can grow the list in the concurrent environment.

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh [hunk 6]
```diff
@@ -822,140 +757,182 @@ void process_and_fill_codes(raft::device_resources const& handle,
...
+                 SizeT new_used_size,
+                 SizeT old_used_size,
+                 uint32_t pq_bits,
+                 uint32_t pq_dim)
+{
+  bool skip_resize = false;
+  if (orig_list) {
+    if (new_used_size <= orig_list->indices.extent(0)) {
+      auto shared_list_size = old_used_size;
+      if (new_used_size <= old_used_size ||
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:836]: I haven't come up with the use case for this, but the function interface allows this. So I figured, I'd support this in case we need it in future :)

File: cpp/include/raft/neighbors/ivf_pq_types.hpp [hunk 7]
```diff
@@ -268,69 +331,44 @@ struct index : ann::index {
...
       codebook_kind_(codebook_kind),
-      n_lists_(n_lists),
       dim_(dim),
       pq_bits_(pq_bits),
       pq_dim_(pq_dim == 0 ? calculate_pq_dim(dim) : pq_dim),
-      n_nonempty_lists_(n_nonempty_lists),
       pq_centers_{make_device_mdarray<float>(handle, make_pq_centers_extents())},
-      pq_dataset_{make_device_mdarray<uint8_t>(handle, make_pq_dataset_extents(0))},
-      indices_{make_device_mdarray<IdxT>(handle, make_extents<IdxT>(0))},
+      lists_{make_host_mdarray<std::shared_ptr<list_data<IdxT>>>(make_extents<uint32_t>(n_lists))},
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/ivf_pq_types.hpp:345]: I'm using `host_vector<std::shared_ptr<list_data<IdxT>>, uint32_t, row_major> lists_` at the moment. Are you looking at the latest commits? :)

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Artem, I have finished my review on the remaining parts, here are few more comments.

File: cpp/include/raft/neighbors/ivf_pq_types.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/ivf_pq_types.hpp]: Ah, ok, why not! Fixed.

File: cpp/include/raft/neighbors/ivf_pq_types.hpp [hunk 8]
```diff
@@ -139,6 +141,79 @@ constexpr static uint32_t kIndexGroupSize = 32;
...
+  /** Source indices. */
+  device_mdarray<IdxT, extent_1d<SizeT>, row_major> indices;
+  /** The actual size of the content. */
+  std::atomic<SizeT> size;
+
+  /** Allocate a new list capable of holding at least `n_rows` data records and indices. */
+  list_data(raft::device_resources const& res, SizeT n_rows, uint32_t pq_bits, uint32_t pq_dim)
+    : size{n_rows}
+  {
+    auto capacity = round_up_safe<SizeT>(bound_by_power_of_two<SizeT>(size), kIndexGroupSize);
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/ivf_pq_types.hpp]: As per our discussion offline, I made the amortizing logic a bit more conservative and added a new boolean 
 to index_params `conservative_memory_allocation` to disable the amortization altogether if needed.

File: cpp/include/raft/neighbors/detail/ivf_pq_serialize.cuh [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+                                          ld->data.extent(2),
+                                          ld->data.extent(3));
+  auto data_array   = make_host_mdarray<uint8_t, SizeT, row_major>(data_extents);
+  auto inds_array   = make_host_mdarray<IdxT, SizeT, row_major>(make_extents<SizeT>(size));
+  deserialize_mdspan(handle, is, data_array.view());
+  deserialize_mdspan(handle, is, inds_array.view());
+  copy(ld->data.data_handle(), data_array.data_handle(), data_array.size(), handle.get_stream());
+  // NB: copying exactly 'size' indices to leave the rest 'kInvalidRecord' intact.
+  copy(ld->indices.data_handle(), inds_array.data_handle(), size, handle.get_stream());
+}
```
[the remark this answers is not in the document]
Comment (tfeher) [reply cpp/include/raft/neighbors/detail/ivf_pq_serialize.cuh:109]: >[cudaMemcpyAsync()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79) is asynchronous with respect to the host, so the call may return before the copy is complete.

You are executing a H2D copy and the host array could go out of scope before the copy is complete. Yes, I think this means we should sync here.

[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/detail/ivf_pq_serialize.cuh:109]: The rmm::device_buffers, which are in the back of our `device_mdarray`, seem to deallocate the memory [asynchronously](https://github.com/rapidsai/rmm/blob/c9c83abb3c5b8285de98aad9c0ebe4f988015029/include/rmm/device_buffer.hpp#L232-L237), so that shouldn't be necessary (otherwise we really need to force sync in many places in raft!).

**Review by tfeher — COMMENTED** [review]
Thanks Artem for the updates! Apart from one discussion point it looks good overall.

One question to @cjnolet: do you have any suggestion for a less verbose name for `conservative_memory_allocation`, or shall we keep that?

File: cpp/include/raft/neighbors/detail/ivf_pq_serialize.cuh (hunk 9 above)
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/detail/ivf_pq_serialize.cuh:109]: Ah, I get in now, you mean the problem is that these two are the host arrays not aware of cuda streams? Good point, I'll add the synchronize and check if I have this pattern anywhere else in the PR.

**Review by tfeher — APPROVED** [review]
Thanks Artem for addressing the issues. The PR looks good to me!

File: cpp/include/raft/neighbors/ivf_pq_types.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/include/raft/neighbors/ivf_pq_types.hpp]: I agree with @tfeher here. Instead of the host vector of shared pointers, I would prefer that we just use std::vector in these cases. Even though a std::vector could be used under the hood anyways, the host_vector is supposed to be treated as a host counterpart to the device_vector- thus used for storing bytes of data.

File: cpp/include/raft/spatial/knn/specializations.cuh [hunk 10]
```diff
@@ -23,6 +23,4 @@
 #include <raft/neighbors/specializations/fused_l2_knn.cuh>
 #include <raft/neighbors/specializations/knn.cuh>
 
-#include <raft/neighbors/specializations/detail/ivf_pq_search.cuh>
```
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/specializations.cuh:26]: Ah, yeah we definitely don't want to include the specializations header for the function we are compiling- that will make it not compile it at all. Good catch!

File: cpp/bench/neighbors/knn.cuh [hunk 11]
```diff
@@ -26,15 +26,12 @@
...
 #if defined RAFT_DISTANCE_COMPILED
 #include <raft/distance/specializations.cuh>
-#include <raft/neighbors/specializations/ivf_pq_build.cuh>
+#include <raft/neighbors/specializations/ivf_pq.cuh>
 #endif
 
 #if defined RAFT_NN_COMPILED
-#include <raft/spatial/knn/specializations.cuh>
-#if defined RAFT_DISTANCE_COMPILED
-#include <raft/cluster/specializations.cuh>
```
Comment (cjnolet) [inline cpp/bench/neighbors/knn.cuh:35]: I had included the clustering specializations here specifically because we are using kmeans within ivfpq. Maybe this is unneeded? Now that the two kmeans impls are sharing a bunch of their core, unless we are instantiating something in ivfpq up the stack from kmeans, we should probably still include this (though it's probably better in the actual knn specializations header if it is actually needed).

File: cpp/include/raft/neighbors/ivf_pq_types.hpp [hunk 12]
```diff
@@ -139,6 +151,90 @@ constexpr static uint32_t kIndexGroupSize = 32;
...
+struct list_data {
+  /** PQ-encoded interleaved data. */
+  device_mdarray<uint8_t, list_extents<SizeT>, row_major> data;
+  /** Source indices. */
+  device_mdarray<IdxT, extent_1d<SizeT>, row_major> indices;
+  /** The actual size of the content. */
+  std::atomic<SizeT> size;
+
+  /** Allocate a new list capable of holding at least `n_rows` data records and indices. */
+  list_data(raft::device_resources const& res,
```
Comment (cjnolet) [inline cpp/include/raft/neighbors/ivf_pq_types.hpp:206]: Aside from the batch insert perf in FAISS, one of the reasons we need this is because we need to support conversion from and to the FAISS cpu and gpu layouts, which is not performant or trivial with a single block of memory. 

The faiss developers have created a class (the CodePacker) to allow us to encapsulate our layout pattern. I think instead of manually implementing the code packer logic for the raft indices in faiss, we should expose a couple little utility functions on our public API (maybe in a raft::neighbors::ivf_pq::helpers namespace like we did with kmeans) that would allow us to specify an index, a code size, a block size, some number of vectors, a chunk of memory with be flat codes, and provide a pointer to an output chunk of memory to write the packed "raft"-encoded vectors to. 

We could also provide another function that would accept the same input arguments but "unpack" a "raft"-encoded block back into its (non interleaved) flat form.

The code layout here is really a deep internal detail of our implementations and I'd prefer that we encapsulate that logic and keep it internal so that we can 

1) keep it close to the indexing code itself, and 

2) make future changes to it as needed  without having to manually change every library that's using raft downstream.

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh [hunk 13]
```diff
@@ -822,140 +757,178 @@ void process_and_fill_codes(raft::device_resources const& handle,
...
 
-/** See raft::spatial::knn::ivf_pq::extend docs */
+/**
+ * Extend the index in-place.
+ * See raft::spatial::knn::ivf_pq::extend docs.
+ */
 template <typename T, typename IdxT>
-auto extend(raft::device_resources const& handle,
-            const index<IdxT>& orig_index,
+void extend(raft::device_resources const& handle,
```
Comment (cjnolet) [inline cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:920]: For the faiss integration, we need to support both extending an index with a new set of vectors as well as being able to write vectors directly into the lists (for the case of converting to/from a faiss index on the cpu/gpu.

It would help if we had a utility function to encode and write the (non-interleaved) flat indices from a block of memory directly to a specific list.

**Review by cjnolet — CHANGES_REQUESTED** [review]
Thanks again for implementing these changes! This is definitely going to benefit the integration into FAISS.

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh [hunk 14]
```diff
@@ -727,6 +662,12 @@ __launch_bounds__(BlockSize) __global__ void process_and_fill_codes_kernel(
   const uint32_t pq_len     = pq_centers.extent(1);
   const uint32_t pq_dim     = new_vectors.extent(1) / pq_len;
 
+  auto pq_extents = make_list_extents<uint32_t>(out_ix + 1, PqBits, pq_dim);
+  auto pq_extents_vectorized =
+    make_extents<uint32_t>(pq_extents.extent(0), pq_extents.extent(1), pq_extents.extent(2));
+  auto pq_dataset = make_mdspan<pq_vec_t, uint32_t, row_major, false, true>(
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:668]: That's a cuda kernel code Corey :) Also this is a really an implementation detail that is hard to workaround.  `pq_vec_t` is in fact a cuda simd type, and the whole extents/mdspan exercise is only needed because we don't have submdspan yet (which I miss a lot!).

File: cpp/bench/neighbors/knn.cuh (hunk 11 above)
Comment (achirkin) [reply cpp/bench/neighbors/knn.cuh:35]: Ah, I forgot that the balanced k-means now lives in these headers as well. Do we already have specializations for it?

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh (hunk 13 above)
Comment (achirkin) [reply cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:920]: Makes sense! I'd suggest we add this in a follow-up PR though. The current implementation is optimized to go over all input data in a single pass, so there's no obvious separation between the sorting input in the clusters and adding them to lists.

File: cpp/include/raft/neighbors/ivf_pq_types.hpp (hunk 12 above)
Comment (achirkin) [reply cpp/include/raft/neighbors/ivf_pq_types.hpp]: List data is a part of the index declaration, so we have to keep it public. Since we're also going to use the same pattern in ivf-flat, I separated it into `ivf_list_types.hpp` and `ivf_list.hpp` and parameterized it with a `SpecT` template, which defines the implementation-specific data layout. I hope this make it more convenient to integrate with the CodePacker later!

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh (hunk 14 above)
[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:669]: > That's a cuda kernel code Corey :) 

Doh! I see that now- the way github presents the diff here somewhat obscures the function signature and I'm more used to seeing the mdspan in non-device code. Yeah this is fine- I was also somewhat responding to Tamas' earlier question as to whether we should prefer to use mdspan<shared_pr<>> or just use std::vector and in those cases I'd prefer to use std::vector.

File: cpp/include/raft/neighbors/ivf_pq_types.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/include/raft/neighbors/ivf_pq_types.hpp]: I still think we can take the matrix fill operation out of this file, but I don't want to hold up these changes from getting merged.

File: cpp/bench/neighbors/knn.cuh (hunk 11 above)
Comment (cjnolet) [reply cpp/bench/neighbors/knn.cuh:35]: I could be mistaken but I believe @Nyrio instantiated the fused 1nn because it was the prime bottleneck in kmeans. Whether or not that instantiation propagates to the balanced algorithm, im not sure.

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh (hunk 13 above)
Comment (cjnolet) [reply cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:872]: Yep I'm definitely okay doing that in a follow up.

File: cpp/include/raft/neighbors/ivf_pq_types.hpp [hunk 15]
```diff
@@ -90,6 +94,16 @@ struct index_params : ann::index_params {
    * regardless of the values of `dim` and `pq_dim`.
    */
   bool force_random_rotation = false;
+  /**
+   * By default, the algorithm allocates more space than necessary for individual clusters
+   * (`list_data`). This allows to amortize the cost of memory allocation and reduce the number of
+   * data copies during repeated calls to `extend` (extending the database).
+   *
+   * The alternative is the conservative allocation behavior; when enabled, the algorithm always
```
Comment (cjnolet) [inline cpp/include/raft/neighbors/ivf_pq_types.hpp:102]: Looks great!

File: cpp/test/CMakeLists.txt (hunk 1 above)
Comment (cjnolet) [reply cpp/test/CMakeLists.txt:250]: Yep, I understand your point here and I'm on board with this. I think we can find a good compromise here- maybe choose an indexing type that we use consistently but have another that can still test. 

As per offline conversation I just wanted to note here that we have been discussing having a set of tests for an expanded set of types that are compiled one nightly to save precious CI time.

**Review by cjnolet — APPROVED** [review]
Thanks again Artem! These changes look great.

File: cpp/include/raft/neighbors/detail/ivf_pq_build.cuh (hunk 14 above)
[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/include/raft/neighbors/detail/ivf_pq_build.cuh:669]: Maybe a sub mdspan is something we should prioritize after things calm down. We have talked about it for awhile.
